extensions/amp-form/0.1/form-validators.js

```diff
@@ -0,0 +1,347 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {ValidationBubble} from './validation-bubble';
+
+
+/** @type {?./validation-bubble.ValidationBubble|undefined} */
+let validationBubble;
+
+
+/** @type {boolean|undefined} */
+let reportValiditySupported;
+
+
+/**
+ * @param {boolean} isSupported
+ * @private visible for testing.
+ */
+export function setReportValiditySupported(isSupported) {
+  reportValiditySupported = isSupported;
+}
+
+
+/** @const @enum {string} */
+const CustomValidationTypes = {
+  AsYouGo: 'as-you-go',
+  ShowAllOnSubmit: 'show-all-on-submit',
+  ShowFirstOnSubmit: 'show-first-on-submit',
+};
+
+
+class FormValidator {
+
+  /**
+   * @param {!HTMLFormElement} form
+   */
+  constructor(form) {
+    /** @const {!HTMLFOrmElement} */
+    this.form = form;
+
+    /** @const {!Document} */
+    this.doc = form.ownerDocument;
+  }
+
+  /**
+   * Called to report validation errors.
+   */
+  report() {}
+
+  /**
+   * @param {!Event} unusedEvent
+   */
+  onBlur(unusedEvent) {}
+
+  /**
+   * @param {!Event} unusedEvent
+   */
+  onInput(unusedEvent) {}
+}
+
+
+class DefaultValidator extends FormValidator {
+
+  /** @override */
+  report() {
+    this.form.reportValidity();
+  }
+
+}
+
+
+class PolyfillDefaultValidator extends FormValidator {
+
+  constructor(form) {
+    super(form);
+
+    if (!validationBubble) {
+      const win = this.doc.defaultView;
+      validationBubble = new ValidationBubble(win, 'amp-validation-bubble');
+    }
+  }
+
+  /** @override */
+  report() {
+    const inputs = this.form.querySelectorAll('input,select,textarea');
+    for (let i = 0; i < inputs.length; i++) {
+      if (!inputs[i].checkValidity()) {
```

Comment:
Do we know the `input` supports `#checkValidity`? We have to [test here](https://github.com/ampproject/amphtml/pull/5006/files#diff-4f7d7ef94609d4d079c3fde230f0d924R380).